File: Problem 21/solution21.cc
#include <iostream>
#include <cmath>

using namespace std;
int d(int);

int main() {
   int sum = 0, i, j, k;
   int limit = 100000;
   for (i = 1; i < limit; i++) {
      j = d(i);
      k = d(j);
      sum += (i != j and i == k) ? i : 0;
   }
   cout << sum << endl;
   return EXIT_SUCCESS;
}

int d(int num) {
   int i, sum = 1, limit = (int)sqrt(num);
   for (i = 2; i <= limit; i++)
      if (num % i == 0)
         sum += (i == num/i) ? i : i + num/i;
   return sum;
}

File: Problem 21/solution21.py
import math

def calcAmicableSum(limit):
   sum = 0
   for i in range(1, limit):
      for j in range(1, limit):
         if (i != j):
            if (d(i) == j) and (d(j) == i):
               sum += i
               print i, "and", j, "-> sum =", sum
   print "Total sum is", sum

def d(num):
   sum, limit = 1, int(math.sqrt(num))
   for i in range(2, limit):
      if (num % i == 0):
         sum += i + num/i
   return sum

